crates/ethereum-rpc/src/ethereum.rs

```diff
@@ -102,6 +108,62 @@ impl<C: sov_modules_api::Context, Da: DaService> Ethereum<C, Da> {
         (U256::from(base_fee), DEFAULT_PRIORITY_FEE)
     }
 
+    pub fn get_proof(
+        &self,
+        address: Address,
+        keys: Vec<U256>,
+        block_id: Option<BlockId>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<reth_rpc_types::EIP1186AccountProofResponse, ErrorObjectOwned> {
+        use sov_state::storage::{StateCodec, StorageKey};
+
+        let evm = Evm::<C>::default();
+        evm.set_state_to_end_of_evm_block_by_block_id(block_id, working_set)?;
+
+        let account = evm.accounts.get(&address, working_set).unwrap_or_default();
+        let balance = account.balance;
+        let nonce = account.nonce;
+        let code_hash = account.code_hash.unwrap_or(KECCAK_EMPTY);
+
+        let account_key = StorageKey::new(
+            evm.accounts.prefix(),
+            &address,
+            evm.accounts.codec().key_codec(),
+        );
+        let account_proof = working_set.get_with_proof(account_key);
+        let account_proof = borsh::to_vec(&account_proof).expect("Serialization shouldn't fail");
+        let account_proof = Bytes::from(account_proof);
+
+        let db_account = DbAccount::new(address);
+        let mut storage_proof = vec![];
+        for key in keys {
+            let storage_key = StorageKey::new(
+                db_account.storage.prefix(),
+                &key,
+                db_account.storage.codec().key_codec(),
+            );
+            let value = db_account.storage.get(&key, working_set);
+            let proof = working_set.get_with_proof(storage_key);
+            let value_proof = borsh::to_vec(&proof.proof).expect("Serialization shouldn't fail");
+            let value_proof = Bytes::from(value_proof);
+            storage_proof.push(EIP1186StorageProof {
+                key: JsonStorageKey(key.into()),
+                value: value.unwrap_or_default(),
+                proof: vec![value_proof],
```

Comment:
this feels weird? why a vec of single length?

I thought the storage proof was a vec of merkle nodes that should be hashed?

Reply:
That's because `let proof: sov_state::storage::StorageProof<<<C as Spec>::Storage as Storage>::Proof> = working_set.get_with_proof(storage_key);` - the real type is generic. I would need to hack into `trait Proof` to properly serialize it with rlp.

Do we need it?

Reply:
we can force the rpc implementation to have NativeStorage = ProverStorage 

so we get a concrete type for the Proof

then we can apply eth-like serialization to the proof